cranelift/docs/isle.md

```diff
@@ -0,0 +1,1306 @@
+*correctness* of the rewrite should not depend on which rule is
+chosen, only possibly the "optimality" of the output (whatever that
+means for the application domain in question) or the number of rewrite
+steps to get there.
+
+However, in order to provide a deterministic answer, the system must
+somehow specify which rule will be applied in such a situation based
+on precedence, or specificity, or some other tie-breaker. A common
+heuristic is "more specific rule wins". We will see how ISLE resolves
+this question below by using both this heuristic and an explicit
+priority mechanism.
+
+#### Right-hand Sides: Rewrite Expressions
+
+Given a rule whose pattern has matched, we now need to compute the
+rewritten term that replaces the original input term. This rewrite is
+specified by the right-hand side (RHS), which consists of an
+*expression* that generates a new term. This expression can use parts
+of the input term that have been captured by variables in the
+pattern.
+
+We have already seen a few examples of this above: simple term
+expressions, with variables used in place of concrete subterms where
+desired. A typical term-rewrite system allows just a few options in
+the output expression:
+
+* Terms, with sub-expressions as arguments;
+* Constant primitives (`42` or `$Symbol`); and
+* Captured variable values (`x`).
+
+The options are more limited in expressions than in patterns (e.g.,
+there are no wildcards) because a pattern is matching on a range of
+possible terms while an expression must specify a specific rewrite
+result.
+
+### Rewrite Steps and Intermediate Terms
+
+Now that we can specify rewrites via a list of rules, we can study how
+the top-level execution of a term-rewriting system proceeds. Much of
+the power of term-rewriting comes from the fact that rewrites can
+*chain together* into a multi-step traversal through several
+intermediate terms before the final answer is computed.
+
+For a simple example, consider the following rules:
+
+```plain
+    (A (B x)) -> (C x)
+    (C (D x)) -> (E x)
+    (C (F x)) -> (G x)
+```
+
+This set of rules will rewrite `(A (B (D 42)))` to `(C (D 42))`, then
+to `(E 42)` (via the first and third rules respectively).
+
+How is this useful? First, rewriting one term to another (here, `C` at
+the top level) that in turn appears in the left-hand side of other
+rules allows us to *factor* a "program" of term-rewriting rules in the
+same way that imperative programs are factored into separate
+functions.[^4] The usual advantages of a well-factored program, where
+each problem is solved with a small step that "reduces to a previously
+solved problem", apply here.
+
+Second, repeating the rewrite step is actually what grants
+term-rewriting its Turing-completeness: it allows for arbitrary
+control flow.[^5] This might be useful in cases where, for example, a
+term-rewriting program needs "loop" over a list of elements in the
+input: it can recurse and use intermediate terms to store state.
```

Comment:
Maybe, but that's IMHO there is value in introducing the philosophy behind the approach and the language as well; I'm thinking a bit broader than just folks who want to hack on Cranelift, and this is meant to be reference documentation more than a tutorial. E.g. someone might not realize that it's possible to solve a problem in a certain way by recursing through term-rewrites, unless one is already familiar with the general field; one might not understand how term-rewriting came about and what it can be used for. I want to be helpful to that sort of newcomer who may not be as well-read in the field :-)

Said another way, somewhere we should do a deep-dive with all the details and thinking behind ISLE; this is meant to be that deep-dive, more than a quickstart.

There is definitely room for a "learn just enough about the DSL to be proficient at instruction lowering rules" tutorial to exist, though; or I guess it already sort of does, in the docs (README's tutorial at least) that you've written?

Reply:
Thinking about this a bit more -- I can probably wrap the whole TRS section in a "This section provides background on the general term-rewriting-system concept; please feel free to skip ahead if you just want details on ISLE" `goto`; that gives the context and background for those who want it without making it a necessary prerequisite.

Reply:
Done as above, added a reader-goto around the section "if you just want details...".